src/native/containers/dn-simdhash-specialization.h

```diff
@@ -0,0 +1,524 @@
+	if (hash->count >= hash->grow_at_count) {
+		// printf ("hash->count %d >= hash->grow_at_count %d\n", hash->count, hash->grow_at_count);
+		return DN_SIMDHASH_INSERT_NEED_TO_GROW;
+	}
+
+	dn_simdhash_buffers_t buffers = hash->buffers;
+	uint8_t suffix = dn_simdhash_select_suffix(key_hash);
+	uint32_t first_bucket_index = dn_simdhash_select_bucket_index(hash->buffers, key_hash);
+	dn_simdhash_suffixes search_vector = build_search_vector(suffix);
+
+	BEGIN_SCAN_BUCKETS(first_bucket_index, bucket_index, bucket_address)
+		// If necessary, check the current bucket for the key
+		if (ensure_not_present) {
+			int index_in_bucket = DN_SIMDHASH_SCAN_BUCKET_INTERNAL(hash, bucket_address, key, search_vector);
+			if (index_in_bucket >= 0)
+				return DN_SIMDHASH_INSERT_KEY_ALREADY_PRESENT;
+		}
+
+		// The current bucket doesn't contain the key, or duplicate checks are disabled (for rehashing),
+		//  so attempt to insert into the bucket
+		uint8_t new_index = dn_simdhash_bucket_count(bucket_address->suffixes);
+		if (new_index < DN_SIMDHASH_BUCKET_CAPACITY) {
+			// Calculate value slot index early so that we don't stall waiting for it later
+			uint32_t value_slot_index = (bucket_index * DN_SIMDHASH_BUCKET_CAPACITY) + new_index;
+			// We found a bucket with space, so claim the first free slot
+			dn_simdhash_bucket_set_count(bucket_address->suffixes, new_index + 1);
+			dn_simdhash_bucket_set_suffix(bucket_address->suffixes, new_index, suffix);
+			bucket_address->keys[new_index] = key;
+			*address_of_value(buffers, value_slot_index) = value;
+			// printf("Inserted [%zd, %zd] in bucket %d at index %d\n", key, value, bucket_index, new_index);
+			// If we cascaded out of our original target bucket, scan through our probe path
+			//  and increase the cascade counters. We have to wait until now to do that, because
+			//  during the process of getting here we may end up finding a duplicate, which would
+			//  leave the cascade counters in a corrupted state
+			adjust_cascaded_counts(buffers, first_bucket_index, bucket_index, 1);
+			return DN_SIMDHASH_INSERT_OK;
+		}
+
+		// The current bucket is full, so try the next bucket.
+	END_SCAN_BUCKETS(first_bucket_index, bucket_index, bucket_address)
+
+	return DN_SIMDHASH_INSERT_NEED_TO_GROW;
+}
+
+static void
+DN_SIMDHASH_REHASH_INTERNAL (DN_SIMDHASH_T_PTR hash, dn_simdhash_buffers_t old_buffers)
+{
+	BEGIN_SCAN_PAIRS(old_buffers, key_address, value_address)
+		uint32_t key_hash = DN_SIMDHASH_KEY_HASHER(DN_SIMDHASH_GET_DATA(hash), *key_address);
+		// This theoretically can't fail, since we just grew the container and we
+		//  wrap around to the beginning when there's a collision in the last bucket.
+		dn_simdhash_insert_result ok = DN_SIMDHASH_TRY_INSERT_INTERNAL(
+			hash, *key_address, key_hash,
+			*value_address,
+			0
+		);
+		// FIXME: Why doesn't assert(ok) work here? Clang says it's unused
+		if (ok != DN_SIMDHASH_INSERT_OK)
+			assert(0);
```

Comment:
maybe `assert` was a no-op in whatever build configuration you were testing?

Reply:
I think so, yeah. I grepped around trying to figure out where in our build configuration we were disabling asserts, and I couldn't, but it seems like that's what is happening.